File: examples/ios/Sources/KandiLoginSDK/KandiLoginClient.swift
// KandiLoginClient.swift
// KandiLoginSDK
//
// Copyright (c) KandiForge. All rights reserved.
// Licensed under the MIT License.

import AuthenticationServices
import Foundation

public final class KandiLoginClient: Sendable {

    // MARK: - Properties

    public let authServerUrl: String
    private let session: URLSession
    private let decoder: JSONDecoder

    // MARK: - Initialization

    public init(authServerUrl: String) {
        self.authServerUrl = authServerUrl
        self.session = URLSession.shared
        self.decoder = JSONDecoder()
    }

    // MARK: - OAuth Login

    /// Opens an ASWebAuthenticationSession to perform OAuth login with the given provider.
    /// Returns a `TokenResponse` extracted from the callback URL.
    @MainActor
    public func login(
        provider: String,
        callbackScheme: String,
        presentationContextProvider: ASWebAuthenticationPresentationContextProviding
    ) async throws -> TokenResponse {
        let loginPath = "/api/auth/login"
        let returnUrl = "\(callbackScheme)://auth/callback"

        guard var components = URLComponents(string: "\(authServerUrl)\(loginPath)") else {
            throw KandiLoginError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "provider", value: provider),
            URLQueryItem(name: "return_url", value: returnUrl),
            URLQueryItem(name: "client_type", value: "desktop"),
        ]

        guard let url = components.url else {
            throw KandiLoginError.invalidURL
        }

        let callbackURL: URL = try await withCheckedThrowingContinuation { continuation in
            let authSession = ASWebAuthenticationSession(
                url: url,
                callbackURLScheme: callbackScheme
            ) { callbackURL, error in
                if let error = error as? ASWebAuthenticationSessionError,
                   error.code == .canceledLogin
                {
                    continuation.resume(throwing: KandiLoginError.authenticationCancelled)
                    return
                }

                if let error {
                    continuation.resume(throwing: KandiLoginError.networkError(error))
                    return
                }

                guard let callbackURL else {
                    continuation.resume(throwing: KandiLoginError.noCallbackURL)
                    return
                }

                continuation.resume(returning: callbackURL)
            }

            authSession.presentationContextProvider = presentationContextProvider
            authSession.prefersEphemeralWebBrowserSession = false
            authSession.start()
        }

        return try parseTokensFromCallback(callbackURL)
    }

    // MARK: - Token Validation

    /// Validates an access token and returns the associated user.
    public func validateToken(_ token: String) async throws -> KandiUser {
        let url = try buildURL(path: "/api/auth/validate")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        return try await performRequest(request)
    }

    // MARK: - Token Refresh

    /// Refreshes an expired access token using a refresh token.
    public func refreshToken(_ refreshToken: String) async throws -> TokenResponse {
        let url = try buildURL(path: "/api/auth/refresh")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["refresh_token": refreshToken]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        return try await performRequest(request)
    }

    // MARK: - Logout

    /// Logs out the current session on the server.
    public func logout(accessToken: String) async throws {
        let url = try buildURL(path: "/api/auth/logout")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (_, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw KandiLoginError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw KandiLoginError.httpError(
                statusCode: httpResponse.statusCode,
                message: "Logout failed"
            )
        }
    }

    // MARK: - Test Endpoints

    /// Seeds test personas on the auth server.
    public func seedTestPersonas() async throws -> SeedResponse {
        let url = try buildURL(path: "/api/auth/test/seed")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        return try await performRequest(request)
    }

    /// Lists all available test personas.
    public func listPersonas() async throws -> PersonasResponse {
        let url = try buildURL(path: "/api/auth/test/personas")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        return try await performRequest(request)
    }

    /// Logs in as a specific test persona by ID.
    public func loginAs(personaId: String) async throws -> TokenResponse {
        let url = try buildURL(path: "/api/auth/test/login-as")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["persona_id": personaId]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        return try await performRequest(request)
    }

    // MARK: - Private Helpers

    private func buildURL(path: String) throws -> URL {
        guard let url = URL(string: "\(authServerUrl)\(path)") else {
            throw KandiLoginError.invalidURL
        }
        return url
    }

    private func performRequest<T: Decodable>(_ request: URLRequest) async throws -> T {
        let data: Data
        let response: URLResponse

        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw KandiLoginError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw KandiLoginError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let message = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw KandiLoginError.httpError(
                statusCode: httpResponse.statusCode,
                message: message
            )
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw KandiLoginError.decodingError(error)
        }
    }

    private func parseTokensFromCallback(_ url: URL) throws -> TokenResponse {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw KandiLoginError.noCallbackURL
        }

        let queryItems = components.queryItems ?? []
        let params = Dictionary(
            uniqueKeysWithValues: queryItems.compactMap { item in
                item.value.map { (item.name, $0) }
            }
        )

        guard let accessToken = params["access_token"],
              let refreshToken = params["refresh_token"]
        else {
            throw KandiLoginError.tokenNotFound
        }

        let expiresIn = params["expires_in"].flatMap { Int($0) } ?? 3600

        return TokenResponse(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresIn: expiresIn
        )
    }
}

File: packages/kandi-login/examples/android/app/src/main/java/com/kandiforge/example/sdk/KandiLoginClient.kt
// Copyright (c) KandiForge. MIT License.

package com.kandiforge.example.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class KandiLoginClient(
    private val authServerUrl: String = "https://kandi-packages-api.vercel.app",
) {

    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json: Json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    // MARK: - OAuth Login URL

    /**
     * Builds the OAuth login URL for the given provider. Open this in a Custom Chrome Tab.
     */
    fun buildLoginUrl(provider: String, callbackScheme: String): String {
        val returnUrl = "$callbackScheme://auth/callback"
        return "$authServerUrl/api/auth/login" +
            "?provider=$provider" +
            "&return_url=${java.net.URLEncoder.encode(returnUrl, "UTF-8")}" +
            "&client_type=desktop"
    }

    /**
     * Parses access and refresh tokens from the callback URI query parameters.
     */
    fun parseTokensFromCallback(uri: android.net.Uri): TokenResponse {
        val accessToken = uri.getQueryParameter("access_token")
            ?: throw KandiLoginException.TokenNotFound
        val refreshToken = uri.getQueryParameter("refresh_token")
            ?: throw KandiLoginException.TokenNotFound
        val expiresIn = uri.getQueryParameter("expires_in")?.toLongOrNull() ?: 3600L

        return TokenResponse(
            accessToken = accessToken,
            refreshToken = refreshToken,
            expiresIn = expiresIn,
        )
    }

    // MARK: - Token Validation

    /**
     * Validates an access token and returns the associated user.
     */
    suspend fun validateToken(token: String): KandiUser = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$authServerUrl/api/auth/validate")
            .get()
            .addHeader("Authorization", "Bearer $token")
            .build()

        executeRequest<KandiUser>(request)
    }

    // MARK: - Token Refresh

    /**
     * Refreshes an expired access token using a refresh token.
     */
    suspend fun refreshToken(refreshToken: String): TokenResponse =
        withContext(Dispatchers.IO) {
            val body = """{"refresh_token":"$refreshToken"}"""
                .toRequestBody(jsonMediaType)

            val request = Request.Builder()
                .url("$authServerUrl/api/auth/refresh")
                .post(body)
                .addHeader("Content-Type", "application/json")
                .build()

            executeRequest<TokenResponse>(request)
        }

    // MARK: - Logout

    /**
     * Logs out the current session on the server.
     */
    suspend fun logout(accessToken: String): Unit = withContext(Dispatchers.IO) {
        val body = "".toRequestBody(jsonMediaType)

        val request = Request.Builder()
            .url("$authServerUrl/api/auth/logout")
            .post(body)
            .addHeader("Authorization", "Bearer $accessToken")
            .build()

        val response = httpClient.newCall(request).execute()

        if (!response.isSuccessful) {
            val message = response.body?.string() ?: "Logout failed"
            throw KandiLoginException.HttpError(response.code, message)
        }
    }

    // MARK: - Test Endpoints

    /**
     * Seeds test personas on the auth server.
     */
    suspend fun seedTestPersonas(): SeedResponse = withContext(Dispatchers.IO) {
        val body = "".toRequestBody(jsonMediaType)

        val request = Request.Builder()
            .url("$authServerUrl/api/auth/test/seed")
            .post(body)
            .addHeader("Content-Type", "application/json")
            .build()

        executeRequest<SeedResponse>(request)
    }

    /**
     * Lists all available test personas.
     */
    suspend fun listPersonas(): PersonasResponse = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$authServerUrl/api/auth/test/personas")
            .get()
            .build()

        executeRequest<PersonasResponse>(request)
    }

    /**
     * Logs in as a specific test persona by ID.
     */
    suspend fun loginAs(personaId: String): TokenResponse = withContext(Dispatchers.IO) {
        val body = """{"persona_id":"$personaId"}"""
            .toRequestBody(jsonMediaType)

        val request = Request.Builder()
            .url("$authServerUrl/api/auth/test/login-as")
            .post(body)
            .addHeader("Content-Type", "application/json")
            .build()

        executeRequest<TokenResponse>(request)
    }

    // MARK: - Private Helpers

    private inline fun <reified T> executeRequest(request: Request): T {
        val response = try {
            httpClient.newCall(request).execute()
        } catch (e: IOException) {
            throw KandiLoginException.NetworkError(e)
        }

        val responseBody = response.body?.string()
            ?: throw KandiLoginException.InvalidResponse

        if (!response.isSuccessful) {
            throw KandiLoginException.HttpError(response.code, responseBody)
        }

        return try {
            json.decodeFromString<T>(responseBody)
        } catch (e: Exception) {
            throw KandiLoginException.DecodingError(e)
        }
    }
}

sealed class KandiLoginException(message: String) : Exception(message) {
    data object InvalidUrl : KandiLoginException("Invalid URL")
    data object InvalidResponse : KandiLoginException("Invalid response from server")
    data object AuthenticationCancelled : KandiLoginException("Authentication was cancelled")
    data object TokenNotFound : KandiLoginException("Token not found in callback")
    class HttpError(val statusCode: Int, val body: String) :
        KandiLoginException("HTTP $statusCode: $body")
    class DecodingError(cause: Exception) :
        KandiLoginException("Failed to decode response: ${cause.message}")
    class NetworkError(cause: IOException) :
        KandiLoginException("Network error: ${cause.message}")
}
